### PR TITLE
Update CRANMirror.CRANMirror to centralize display logic

### PR DESCRIPTION
**Pull request by ca-scribner**
### Intent

> Describe briefly what problem this pull request resolves, or what new feature it introduces. Include screenshots of any new or altered UI. Link to any Github issues that are related. 

Update to CRANMirror that makes logic around how to display it to users, as well as whether it is a standard CRAN Mirror or if it is user defined by URL.  Change should not affect function, only make things more maintainable.  This was done while working on internationalizing RStudio and trying to simplify where hardcoded strings are located in the code.

### Approach

> Describe the approach taken and the tradeoffs involved if non-obvious; add an overview of the solution if it's complicated.

CRANMirror implements a .getDisplay() method for showing a human readable representation of itself.  PackagesPreferencesPane.java and ChooseMirrorDialog.java both use this, but only when CRANMirror.host != "Custom".  CRANMirror.host is treated as an enumeration when =="Custom".  When CRANMirror.host == "Custom", they both show CRANMirror.url instead of the host and name. 
* These two references are the only usages of getDisplay(), and are direct copies of each other.  Both apply the same logic in whether to use getDisplay() or create their own displayed text, and when creating their own displayed text they produce the same result (using the URL as the display value
* Determining whether a CRANMirror is "Custom" depends on the somewhat brittle fact that .host=="Custom" when a mirror is defined by URL alone.  This is hard for the CRANMirror maintain if it was ever updated (need to find hard coded strings of "Custom" in the codebase and update them)

Proposed changes here address the above by centralizing the display logic of PackagesPreferencesPane.java and ChooseMirrorDialog.java in CRANMirror.  The following added/changed methods are provided in CRANMirror:
* .getCustomEnumValue(): Serves as an enumerator to specify the string label that denotes a custom object.  This would be more cleanly implemented as an enumerator string, but CRANMirror extends JavaScriptObject which cannot have class-level attributes
* .getSecondaryEnumValue(): Similar to Custom, but implements another enumerator for a "Secondary" mirror
* .setAsCustom(): This sets the object as a "Custom" (url-only) object, replacing the previous pattern where the user of the object would set .host="Custom" and .name="Custom" themselves
* .isCustom(): API for public to assess whether a CRANMirror is or is not "Custom".  Implemented by testing `host==getCustomEnumValue()` (eg: host=="Custom"), which is how consumers of this object previously performed the test themselves
* .getDisplay(): Implements the logic used by PackagesPreferencesPane.java and ChooseMirrorDialog.java, where the display text will be the URL if this is a custom CRANMirror.

### Automated Tests

> Indicate whether this change includes unit tests or integration tests, or link a work item or pull request to add those tests to another repo. If the change cannot or will not be covered by a test, indicate why.

No addition to the test suite made.  Function of the code should be the same as prior to the change

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues. 

Can see this working in Global Options -> Packages -> display text for CRAN Mirror

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

Contributor agreement signed and sending later today

### CODE REVIEW

File: src/gwt/src/org/rstudio/studio/client/workbench/prefs/views/PackagesPreferencesPane.java [hunk 1]
```diff
@@ -364,6 +348,7 @@ public RestartRequirement onApply(UserPrefs prefs)
       String mirrorTextValue = cranMirrorTextBox_.getTextBox().getText();
 
       boolean cranRepoChanged = !mirrorTextValue.equals(cranMirrorStored_);
+      // TODO: Is this correct?  Custom doesn't require a leading http://, so this wont always match
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/prefs/views/PackagesPreferencesPane.java:351]: We don't usually leave TODOs in the code, although you're right that there could be an issue here (please file it separately!)

File: src/gwt/src/org/rstudio/studio/client/common/mirrors/model/CRANMirror.java [hunk 2]
```diff
@@ -111,8 +111,48 @@ public final void setSecondaryRepos(ArrayList<CRANMirror> repos)
...
+      } else {
+         return getName() + " - " + getHost();
+      }
+   }
+
+   // Implement enumerators as functions because this extends a JavaScriptObject
+   // A cleaner way might be to wrap this in a wrapper class, but this gets the job done
+   public final static String getCustomEnumValue()
+   {
+      return "Custom"; //$NON-NLS-1$
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/common/mirrors/model/CRANMirror.java:150]: What do the `NON-NLS-1` comments mean?

File: src/gwt/src/org/rstudio/studio/client/common/mirrors/model/CRANMirror.java [hunk 3]
```diff
@@ -111,8 +111,48 @@ public final void setSecondaryRepos(ArrayList<CRANMirror> repos)
...
+
+   /**
+    * Returns a formatted display name for this CRANMirror.
+    *
+    * Returned name includes Name and Host if a standard host, and simply the URL of a custom host from the user
+    */
    public final String getDisplay()
    {
-      return getName() + " - " + getHost();
+      if (isCustom()) {
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/common/mirrors/model/CRANMirror.java:138]: Nit: Code style standard calls for opening brace on its own line.

**Review by jmcphers — COMMENTED** [review]
Looks pretty straightforward modulo a few questions, but I don't think I understand the intent of this change other than to clean up the logic. Does this fix any bugs you're aware of? Is it a prerequisite to i8n work?

Also, we haven't received a CLA yet -- did you send it?

File: src/gwt/src/org/rstudio/studio/client/common/mirrors/model/CRANMirror.java (hunk 2 above)
Comment (ca-scribner) [reply src/gwt/src/org/rstudio/studio/client/common/mirrors/model/CRANMirror.java:151]: This is an indicator to most IDEs (at least Eclipse and IntelliJ) to ignore hard coded string literals on this line (eg: literals that should not be internationalized)

**Comment by ca-scribner** [comment]
@jmcphers thanks for the good feedback!  Sorry for the slow response, I got side tracked on something else.

Both nits above are now resolved.  Re intent: this was an admittedly minor thing that came up in other i18n efforts.  It makes things a tiny bit easier to internationalize (mainly by reducing the number of hardcoded string literals).  The intent is to not change actual function seen by the user

**Comment by ca-scribner** [comment]
And yes you were correct about the contributor agreement - it completely slipped my mind that evening.  I've now sent it.

**Comment by jmcphers** [comment]
Thanks for sending the CLA, this will be merged shortly!

**Pull request merged**